src/AspNet.Security.OAuth.Introspection/Events/ValidateAudienceContext.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace AspNet.Security.OAuth.Introspection {
+    public class ValidateAudienceContext : BaseIntrospectionContext {
+        public ValidateAudienceContext(
+            HttpContext context, 
+            OAuthIntrospectionOptions options,
+            JToken payload) 
+            : base(context, options) {
+            Payload = payload;
+            var audiences = payload[OAuthIntrospectionConstants.Claims.Audience];
+            if(audiences != null) {
+                if(audiences.Type == JTokenType.String) {
+                    Audiences = new[] { payload.Value<string>(OAuthIntrospectionConstants.Claims.Audience) };
+                }
+                else if (audiences.Type == JTokenType.Array) {
+                    Audiences = payload.Value<JArray>(OAuthIntrospectionConstants.Claims.Audience)
+                                           .Select(audience => audience.Value<string>());
+                }
+            }
+            Validate();
+        }
+
+        public IEnumerable<string> Audiences { get; set; }
+
+        public JToken Payload { get; }
+
+        public bool IsValid { get; set; }
+
+        public bool Validate() {
+            return IsValid = Validate(Audiences);
+        }
+
+        public bool Validate(IEnumerable<string> audiences)
+        {
+            IsValid = (audiences == null && Options.Audiences.Count > 0) || Options.Audiences.Count == 0 ||
```

Comment:
True, but ASOS accepts a `client_id`/`redirect_uri` because the dev' is able to override the ones extracted from the authorization request and use them for the rest of the request processing. Here, it doesn't alter anything.

I'll think a bit more about that.